### ISSUE TITLE
Accurate metrics cannot always be computed with batch wise averages.

### ISSUE THREAD

**Issue by brge17** [keras-team/keras#8410]
Keras computes batch wise averages of metrics. This is insufficient for metrics like AUC and hence the legacy metrics were deprecated. Many of issues have been raised on this subject.

As a work around I have a custom callback that takes a generator, a metric function, and a name for the output metric. At the end of a train epoch, the custom callback uses the generator (which is the validation generator), yields the data (X) + label (y), predicts the output on the validation set, computes the metric function on the entire predictions with the labels, and hacks the metric to be written to the logs where it can be subsequently used by model checkpoint, early stopping, etc... It is fully functional but requires the user to disable the normal validation generator.

**Issue 1:**
The evaluation generator function should have a means of doing this automatically, there ideally would be a simple flag that gives the option to have metrics computed at the end instead of each batch.

**Issue Two:**

The workaround, having a for loop with predict on batch has significantly worse performance than predict on generator (1.3-1.4x slower):

```
	for ind, batch in enumerate(self.gen_handle.generate()):

		# Grab batch components
		X_batch = batch[0]
		y_true_batch = batch[1]

		# Compute the output on the batch
		y_pred_batch = self.model.predict_on_batch(X_batch)
```

This issue could be avoided if there was an option to do predict with generator that has a passthrough for labels (or extend evaluate_generator to return y_pred, y_true).

```
y_pred, y_true = self.model.predict_generator(..., passthrough=True)

```

**Comment by ophiry** [comment]
@brge17 can you share the code that updates the logs?

> hacks the metric to be written to the logs where it can be subsequently used by model checkpoint, early stopping, etc...

**Comment by brge17** [comment]
@ophiry Sure.

```
 logs[metric_name] = np.float64(metric_score)
```
np.float64 is for TensorBoard. If you order the callbacks properly, {TrueMetrics, ModelCheckpoint, Earlystopping, etc...} you can have them monitor metric_name.

**Comment by ophiry** [comment]
Thanks, tried a similar thing, but forgot to handle the callback order
